shell/platform/android/io/flutter/plugin/common/MessageCodec.java

```diff
@@ -13,6 +13,16 @@
  * <p>Both operations throw {@link IllegalArgumentException}, if conversion fails.
  */
 public interface MessageCodec<T> {
+
+  /**
+   * Controls the ByteBuffer parameter for `decodeMessage`.
+   *
+   * @see MessageCodec.decodeMessage
+   * @return true if the MessageCodec wants the ByteBuffer parameter to decodeMessage to be a direct
+   *     ByteBuffer.
+   */
+  boolean wantsDirectByteBufferForDecoding();
```

Comment:
So, there were 2 places that the direct ByteBuffers could break someone: MessageCodec subclasses and BinaryCodec users.  We could make sure that the latter is fine.  If we don't have this method, a MessageCodec implementor that stores ByteBuffers will all of a sudden get runtime errors instead of failing to compile because they didn't implement this method.

Are we fine saying that they will get runtime exceptions now, or should we request they implement this method to be sure?  The work around would be simple to implement in the off chance that anyone ever did that.

You think that's fine?  I'm fine with it.  If we aren't willing to make changes like this we will accumulate tons of switches and knobs like this as Flutter ages.

Reply:
A runtime exception that is associated with a `ByteBuffer` method call is a reasonable way to signal this error.

What I'm worried about is anything that could cause segfaults - specifically a `ByteBuffer` backed by a native buffer that is still referenced after the native buffer is freed.  At this point I'm convinced that resetting the buffer's limit should be enough to prevent most usages of `ByteBuffer` from accessing the invalid native buffer and potentially crashing.